src/sinks/loki/event.rs

```diff
@@ -22,8 +28,29 @@ impl Encoder<Vec<LokiRecord>> for LokiBatchEncoder {
         writer: &mut dyn io::Write,
     ) -> io::Result<usize> {
         let batch = LokiBatch::from(input);
-        let body = serde_json::json!({ "streams": [batch] });
-        let body = serde_json::to_vec(&body)?;
+        let body = match self.0 {
+            LokiBatchEncoding::Json => {
+                let body = serde_json::json!({ "streams": [batch] });
+                let body = serde_json::to_vec(&body)?;
+                body
+            }
+            LokiBatchEncoding::Protobuf => {
+                let labels = batch.stream;
+                let entries = batch
+                    .values
+                    .iter()
+                    .map(|event| {
+                        loki_logproto::util::Entry(
+                            event.timestamp,
+                            String::from_utf8_lossy(&event.event).into_owned(),
+                        )
+                        .into()
+                    })
+                    .collect();
+                let batch = loki_logproto::util::Batch(labels, entries);
+                batch.encode()
+            }
+        };
```

Comment:
In reference to my comment about the frame encoder: I realize you're just modifying the existing code, which is totally fine... but we should switch this to write directly the `&mut dyn io::Write` object rather than using an intermediate `Vec<u8>`.

Reply:
As comment above, Loki accept snappy-compressed data in raw format so far. It's unavoidable to meet an intermediate `Vec<u8>`. 😔

Reply:
Yep, fair enough. 👍🏻